extensions/amp-render/1.0/amp-render.js

```diff
@@ -0,0 +1,188 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {BaseElement} from './base-element';
+import {Services} from '../../../src/services';
+import {batchFetchJsonFor} from '../../../src/batched-json';
+import {dev, user, userAssert} from '../../../src/log';
+import {dict} from '../../../src/utils/object';
+import {isExperimentOn} from '../../../src/experiments';
+
+/** @const {string} */
+const TAG = 'amp-render';
+
+const AMP_STATE_URI_SCHEME = 'amp-state:';
+const AMP_SCRIPT_URI_SCHEME = 'amp-script:';
+
+/**
+ * Returns true if element's src points to amp-state.
+ * @param {?string} src
+ * @return {boolean}
+ */
+const isAmpStateSrc = (src) => src && src.startsWith(AMP_STATE_URI_SCHEME);
+
+/**
+ * Returns true if element's src points to an amp-script function.
+ * @param {?string} src
+ * @return {boolean}
+ */
+const isAmpScriptSrc = (src) => src && src.startsWith(AMP_SCRIPT_URI_SCHEME);
+
+class AmpRender extends BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?../../../src/service/template-impl.Templates} */
+    this.templates_ = null;
+
+    /** @private {?Element} */
+    this.template_ = null;
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    userAssert(
+      isExperimentOn(this.win, 'amp-render'),
+      'Experiment "amp-render" is not turned on.'
+    );
+    return super.isLayoutSupported(layout);
+  }
+
+  /** @override */
+  init() {
+    return dict({
+      'getJson': this.getJsonFn_(),
+    });
+  }
+
+  /**
+   * Returns the correct fetch function for amp-state, amp-script or
+   * to fetch remote JSON.
+   *
+   * @return {Function}
+   * @private
+   */
+  getJsonFn_() {
+    const src = this.element.getAttribute('src');
+    if (!src) {
+      // TODO(dmanek): assert that src is provided instead of silently failing below.
+      return () => {};
+    }
+    if (isAmpStateSrc(src)) {
+      return this.getAmpStateJson.bind(null, this.element);
+    }
+    if (isAmpScriptSrc(src)) {
+      // TODO(dmanek): implement this
+      return () => {};
+    }
+    return batchFetchJsonFor.bind(null, this.getAmpDoc(), this.element);
+  }
+
+  /**
+   * TODO: this implementation is identical to one in amp-data-display &
+   * amp-date-countdown. Move it to a common file and import it.
+   *
+   * @override
+   */
+  checkPropsPostMutations() {
+    const templates =
+      this.templates_ ||
+      (this.templates_ = Services.templatesForDoc(this.element));
+    const template = templates.maybeFindTemplate(this.element);
+    if (template != this.template_) {
```

Comment:
super nitty: if you invert the condition via `if (template === this.template_)`, then you can early return and unindent the rest of the fn.